File: sdks/python/dcp_ai/v2/crypto_provider.py
"""
Abstract base classes for pluggable cryptographic providers (signing and KEM).
"""

from __future__ import annotations

import hashlib
from abc import ABC, abstractmethod


def derive_kid(alg: str, public_key_bytes: bytes) -> str:
    """Derive a key ID from algorithm name and raw public key bytes.

    kid = hex(SHA-256(UTF8(alg) || 0x00 || raw_public_key_bytes))[0:32]
    """
    digest = hashlib.sha256(alg.encode("utf-8") + b"\x00" + public_key_bytes).digest()
    return digest[:16].hex()


class CryptoProvider(ABC):
    """Abstract signature provider supporting classical and post-quantum algorithms."""

    alg: str
    key_size: int
    sig_size: int
    is_constant_time: bool

    @abstractmethod
    async def generate_keypair(self) -> dict[str, str]:
        """Generate a keypair. Returns dict with kid, public_key_b64, secret_key_b64."""
        ...

    @abstractmethod
    async def sign(self, message: bytes, secret_key_b64: str) -> bytes:
        """Sign a message. Returns raw signature bytes."""
        ...

    @abstractmethod
    async def verify(self, message: bytes, signature: bytes, public_key_b64: str) -> bool:
        """Verify a detached signature on a message."""
        ...


class KemProvider(ABC):
    """Abstract key encapsulation mechanism provider."""

    alg: str

    @abstractmethod
    async def generate_keypair(self) -> dict[str, str]:
        """Generate a KEM keypair. Returns dict with public_key_b64, secret_key_b64."""
        ...

    @abstractmethod
    async def encapsulate(self, public_key_b64: str) -> dict[str, str]:
        """Encapsulate against a public key. Returns dict with shared_secret, ciphertext_b64."""
        ...

    @abstractmethod
    async def decapsulate(self, ciphertext_b64: str, secret_key_b64: str) -> bytes:
        """Decapsulate a ciphertext using the secret key. Returns shared secret bytes."""
        ...

File: sdks/rust/src/v2/crypto_provider.rs
use sha2::{Sha256, Digest};

#[derive(Debug, thiserror::Error)]
pub enum CryptoError {
    #[error("Key error: {0}")]
    KeyError(String),
    #[error("Signature error: {0}")]
    SignatureError(String),
    #[error("Unknown algorithm: {0}")]
    UnknownAlgorithm(String),
}

#[derive(Debug, Clone)]
pub struct GeneratedKeypair {
    pub kid: String,
    pub public_key_b64: String,
    pub secret_key_b64: String,
}

pub trait CryptoProvider: Send + Sync {
    fn alg(&self) -> &str;
    fn key_size(&self) -> usize;
    fn sig_size(&self) -> usize;
    fn is_constant_time(&self) -> bool;

    fn generate_keypair(&self) -> Result<GeneratedKeypair, CryptoError>;
    fn sign(&self, message: &[u8], secret_key_b64: &str) -> Result<Vec<u8>, CryptoError>;
    fn verify(&self, message: &[u8], signature: &[u8], public_key_b64: &str) -> Result<bool, CryptoError>;
}

/// Key Encapsulation Mechanism provider trait (for KEM algorithms like ML-KEM-768).
pub trait KemProvider: Send + Sync {
    fn alg(&self) -> &str;
    fn pk_size(&self) -> usize;
    fn sk_size(&self) -> usize;
    fn ct_size(&self) -> usize;
    fn ss_size(&self) -> usize;

    fn generate_keypair(&self) -> Result<GeneratedKeypair, CryptoError>;
    fn encapsulate(&self, public_key_b64: &str) -> Result<(Vec<u8>, Vec<u8>), CryptoError>;
    fn decapsulate(&self, ciphertext: &[u8], secret_key_b64: &str) -> Result<Vec<u8>, CryptoError>;
}

/// Derive a key ID from algorithm name and raw public key bytes.
/// kid = hex(SHA-256(UTF8(alg) || 0x00 || public_key_bytes))[0..32]
pub fn derive_kid(alg: &str, public_key_bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(alg.as_bytes());
    hasher.update([0x00]);
    hasher.update(public_key_bytes);
    let hash = hex::encode(hasher.finalize());
    hash[..32].to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_derive_kid_length() {
        let kid = derive_kid("ed25519", &[1, 2, 3]);
        assert_eq!(kid.len(), 32);
    }
}
